src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -1188,22 +1211,24 @@ void addSubpatternsForTuple(ImmutableArray<TypeWithAnnotations> elementTypes)
             bool hasErrors,
             DiagnosticBag diagnostics)
         {
-            // PROTOTYPE(ngafter): prevent designators under 'or' and 'not'
-            var subPattern = BindPattern(node.Pattern, inputType, inputValEscape, hasErrors, diagnostics);
+            const bool permitDesignations = false; // prevent designators under 'not'
+            var subPattern = BindPattern(node.Pattern, inputType, inputValEscape, permitDesignations, hasErrors, diagnostics);
             return new BoundNegatedPattern(node, subPattern, inputType, hasErrors);
         }
 
         private BoundPattern BindBinaryPattern(
             BinaryPatternSyntax node,
             TypeSymbol inputType,
             uint inputValEscape,
+            bool permitDesignations,
             bool hasErrors,
             DiagnosticBag diagnostics)
         {
-            // PROTOTYPE(ngafter): prevent designators under 'or' and 'not'
-            var left = BindPattern(node.LeftPattern, inputType, inputValEscape, hasErrors, diagnostics);
-            var right = BindPattern(node.RightPattern, inputType, inputValEscape, hasErrors, diagnostics);
-            return new BoundBinaryPattern(node, disjunction: node.PatternOperator.Kind() == SyntaxKind.OrKeyword, left, right, inputType, hasErrors);
+            bool isDisjunction = node.PatternOperator.Kind() == SyntaxKind.OrKeyword;
+            permitDesignations = permitDesignations & !isDisjunction; // prevent designators under 'or'
```

Comment:
Any reason for single-and instead of double-and here?

Reply:
Yes.  The straight-line code using locals is more efficient than control-flow.